lib/checkunusedvar.cpp

```diff
@@ -1489,8 +1494,29 @@ bool CheckUnusedVar::isRecordTypeWithoutSideEffects(const Type* type)
         if (f.argDef && Token::simpleMatch(f.argDef->link(), ") ="))
             continue; // ignore default/deleted constructors
         const bool emptyBody = (f.functionScope && Token::simpleMatch(f.functionScope->bodyStart, "{ }"));
-        const bool hasInitList = f.argDef && Token::simpleMatch(f.argDef->link(), ") :");
-        if (!emptyBody || hasInitList)
+
+        Token* nextToken = f.argDef->link();
+        if (Token::simpleMatch(nextToken, ") :")) {
+            // validating initialization list
+            nextToken = nextToken->next(); // goto ":"
+
+            for (const Token *initListToken = nextToken; Token::Match(initListToken, "[:,] %var% [({]"); initListToken = initListToken->linkAt(2)->next())
+            {
+                const Token* var_token = initListToken->next();
+                const Variable* variable = var_token->variable();
+                if (variable && !isVariableWithoutSideEffects(*variable))
+                {
+                    return withoutSideEffects = false;
+                }
+                const Variable* init_value_var = var_token->tokAt(2)->variable();
```

Comment:
I think a loop would be better.

For `x(y)` .. it seems unlikely that there will be a side effect if we just take the value of `y` but not entirely technically impossible. If x and y are pointers or if x is a reference then I have the feeling that side effects there will be impossible. I'll leave the decision to you though, I can test later if that is better. If a non-const method is called like `x(y.dostuff())` then side effects are pretty likely.

I would like that no warning is written here:

    struct Fred {
        Fred() : x(sideEffect()) {}
        int x;
    };
    void foo() {
        Fred fred; // <- do not warn about fred.
    }

Reply:
Sorry, didn't understand you fully from first time.
Your test passed, but I'm not sure that following condition is correct:
```
if ((valueToken->tokType() == Token::Type::eFunction) || 
    (valueToken->tokType() == Token::Type::eName) ||
    (valueToken->tokType() == Token::Type::eLambda) ||
    (valueToken->tokType() == Token::Type::eOther))
{
    return withoutSideEffects = false;
}
```